Dialog for allowing the user to choose the change output when bumping a tx

achow101 opened the pull request:
Based on https://github.com/bitcoin/bitcoin/pull/26467

Implements a GUI dialog for allowing the user to choose the output to reduce when bumping a transaction. This adds the functionality that was added to the RPC.

achow101 committed:
interfaces: Expose CreateRateBumpTransaction's orig_change_pos

achow101 committed:
interfaces: Expose CreateRateBumpTransaction's orig_change_pos

achow101 committed:
interfaces: Add isChange to wallet interface

achow101 committed:
interfaces: Add isChange to wallet interface

achow101 committed:
qt: Add a dialog to select the change output when bumping fee

In order to correctly choose the change output when doing fee bumping in
the GUI, we need to ask the user which output is change. We can make a
guess using our ScriptIsChange heuristic, however the user may have
chosen to have a custom change address or have otherwise labeled their
change address which makes our change detection fail. By asking the user
when fee bumping, we can avoid adding additional change outputs that are
unnecessary.

achow101 committed:
qt: Add a dialog to select the change output when bumping fee

In order to correctly choose the change output when doing fee bumping in
the GUI, we need to ask the user which output is change. We can make a
guess using our ScriptIsChange heuristic, however the user may have
chosen to have a custom change address or have otherwise labeled their
change address which makes our change detection fail. By asking the user
when fee bumping, we can avoid adding additional change outputs that are
unnecessary.

Sjors commented:
Can you add a screenshot to the PR description?

john-moffett reviewed (commented):
Approach ACK

I think I'd prefer if the user had some more information about the outputs in the dialog. Maybe something like this, but worded better:

<img width="729" alt="image" src="https://user-images.githubusercontent.com/116917595/214340781-62f1105a-a5b8-41ae-b96e-ac63e5207fc6.png">
 
Maybe the actual change _address_ isn't even needed, or could be revealed in a tooltip. 

Also, if there's a single change output with sufficient funds for feebumping, then selecting "none" has basically no effect compared to selecting that output, as the wallet will always(?) deduct from that existing change output. Maybe this common case can be simplified by removing the "None" option?

Sjors commented:
Can we be a bit more bold and just assume that a change address is change, and not show the dialog in that case? Ideally we don't show such a dialog more often than necessary.

john-moffett commented:
That would prevent the case where you wanted to deduct the extra fee from the recipient, such as when you initially selected "Subtract fee from amount". 

Ideally, you'd only show the dialog if that were selected previously, but I don't _think_ it's persisted anywhere. Eg -

https://github.com/bitcoin-core/gui/blob/30f553d45797847166109a161628dba3bf00bc94/src/qt/sendcoinsrecipient.h#L42

I could be wrong, though.

luke-jr commented:
Concept ACK. I'd just show "Change" for the change, though. End users shouldn't need to know about change addresses.

hebasto commented on src/qt/bumpfeechoosechangedialog.cpp:
```diff
@@ -0,0 +1,75 @@
+                address_info = QString::fromStdString(address);
+            }
+        }
+        QString output_info = QString::number(i) + QString(": ") + BitcoinUnits::formatWithUnit(model->getOptionsModel()->getDisplayUnit(), txout.nValue) + tr(" to ") + address_info;
```
To make translators' life easier and to keep the context as large as possible, I suggest to use `QString::arg`. Also it will fix the `lint-qt-translation.py` warning.

achow101 replied:
Done

hebasto reviewed (commented):
The first commit e1ea0ba0f2aa9b36df6b74b664087b236f174990 fails to compile:
```
  CXX      qt/libbitcoinqt_a-walletmodel.o
qt/walletmodel.cpp: In member function ‘bool WalletModel::bumpFee(uint256, uint256&)’:
qt/walletmodel.cpp:486:41: error: no matching function for call to ‘interfaces::Wallet::createBumpTransaction(uint256&, wallet::CCoinControl&, std::vector<bilingual_str>&, CAmount&, CAmount&, CMutableTransaction&)’
  486 |     if (!m_wallet->createBumpTransaction(hash, coin_control, errors, old_fee, new_fee, mtx)) {
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./qt/walletmodel.h:16,
                 from qt/walletmodel.cpp:9:
./interfaces/wallet.h:166:18: note: candidate: ‘virtual bool interfaces::Wallet::createBumpTransaction(const uint256&, const wallet::CCoinControl&, std::vector<bilingual_str>&, CAmount&, CAmount&, CMutableTransaction&, std::optional<unsigned int>)’
  166 |     virtual bool createBumpTransaction(const uint256& txid,
      |                  ^~~~~~~~~~~~~~~~~~~~~
./interfaces/wallet.h:166:18: note:   candidate expects 7 arguments, 6 provided
make: *** [Makefile:13550: qt/libbitcoinqt_a-walletmodel.o] Error 1
```

achow101 commented:
Rebased and fixed the build issue in the first commit.

w0xlt reviewed (commented):
Approach ACK

hebasto commented:
@achow101 

`test_bitcoin-qt` still [fails](https://github.com/bitcoin-core/gui/actions/runs/25048051241/job/73368281475) in the CI on Windows.

achow101 commented:
> `test_bitcoin-qt` still [fails](https://github.com/bitcoin-core/gui/actions/runs/25048051241/job/73368281475) in the CI on Windows.

Huh, it works when run directly but not in ctest.

achow101 committed:
gui, test: Ensure buttons are actually pressed before returning

ConfirmSend and ConfirmMessage assumed that the widgets with the buttons
to click would be ready by the time the functions fired. This may not
necessarily be the case, so instead they should check for the right
widgets in a loop and only exit when a button has been clicked.

pablomartin4btc reviewed (commented):
Tracked down the CI issue. The deadlock is in the test helpers: `ConfirmSend` and `ChooseBumpfeeOutput` used `while(true)` spin loops inside `QTimer::singleShot` callbacks to wait for dialogs to appear. Spinning inside the callback blocks the Qt event loop, so `QDialog::exec()` can never make the dialog visible.

The fix replaces the spin loops with a non-blocking retry: if the dialog is not yet visible, reschedule with a 10 ms delay so the next check fires in a later event loop iteration.

Branch with fix on top: [pablomartin4btc:pr700-rebased](https://github.com/pablomartin4btc/bitcoin-core-gui/tree/pr700-rebased) and CI green: [run 30593340908](https://github.com/pablomartin4btc/bitcoin-core-gui/actions/runs/30593340908).

I'll be opening a rebased PR with this fix integrated.
